### ISSUE TITLE
Port rule CA1011 - roslyn-analyzers

### ISSUE THREAD

**Issue by Evangelink** [dotnet/runtime#61916]
I think we should file a separate issue on dotnet\runtime and allow the runtime team to triage the correct severity/enabled by default state of this ported FxCop rule. @buyaa-n @carlossanlop

_Originally posted by @mavasani in https://github.com/dotnet/roslyn-analyzers/pull/5356#discussion_r735440377_

**Comment by carlossanlop** [comment]
The usage is pretty straightforward.

The original issue is [here](https://github.com/dotnet/roslyn-analyzers/issues/375).

There is already a [PR porting this fixer from FxCop](https://github.com/dotnet/roslyn-analyzers/pull/5356), and it's blocked by having this proposal approved (or not).

Existing MS Docs for CA1011: https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1011?view=vs-2022

- Suggested category: Design (same as FxCop)
- Suggested severity: Hidden, because it's currently labeled as an analyzer/fixer that can cause a breaking change, and could potentially be noisy.
- Suggested milestone: Future

### Flag
```cs
        // Before
        public void ReadNextByte(FileStream stream)
        {
            while (stream.ReadByte() != -1) { /*...*/ }
        }

        // After
        public void ReadNextByte(Stream anyStream)
        {
            while (anyStream.ReadByte() != -1) { /*...*/ }
        }
```

### Do not flag
```cs
        // Before
        public void ReadHandle(FileStream stream)
        {
            while (anyStream.ReadByte() != -1) { /*...*/ } // ReadByte is available in Stream
            var handle = stream.SafeFileHandle; // But SafeFileHandle is only available in FileStream
        }
```

### Do not flag
```cs
interface IResource
{
    string Location { get; }
}

class FileResource : IResource
{
    // Location represents a file path
}
class HttpResource : IResource
{
    // Location represents an HTTP url
}

// Should not be corrected to IResource
string GetContent(FileResource res)
{
    return File.ReadAllText(res.Location); // the type encodes information about what the object represents
}
```

**Comment by bartonjs** [comment]
* Generally looks good as proposed, but we changed Hidden to IdeSuggestion

Category: Design
Severity: IdeSuggestion

**Comment by carlossanlop** [comment]
PR implementing this: https://github.com/dotnet/roslyn-analyzers/pull/5356
